### PR TITLE
[SPARK-32753][SQL] Only copy tags to node with no tags

### PR DESCRIPTION
**Pull request by manuzhang**
### What changes were proposed in this pull request?
Only copy tags to node with no tags when transforming plans.

### Why are the changes needed?
@cloud-fan [made a good point](https://github.com/apache/spark/pull/29593#discussion_r482013121) that it doesn't make sense to append tags to existing nodes when nodes are removed. That will cause such bugs as duplicate rows when deduplicating and repartitioning by the same column with AQE.

```
spark.range(10).union(spark.range(10)).createOrReplaceTempView("v1")
val df = spark.sql("select id from v1 group by id distribute by id") 
println(df.collect().toArray.mkString(","))
println(df.queryExecution.executedPlan)

// With AQE
[4],[0],[3],[2],[1],[7],[6],[8],[5],[9],[4],[0],[3],[2],[1],[7],[6],[8],[5],[9]
AdaptiveSparkPlan(isFinalPlan=true)
+- CustomShuffleReader local
   +- ShuffleQueryStage 0
      +- Exchange hashpartitioning(id#183L, 10), true
         +- *(3) HashAggregate(keys=[id#183L], functions=[], output=[id#183L])
            +- Union
               :- *(1) Range (0, 10, step=1, splits=2)
               +- *(2) Range (0, 10, step=1, splits=2)

// Without AQE
[4],[7],[0],[6],[8],[3],[2],[5],[1],[9]
*(4) HashAggregate(keys=[id#206L], functions=[], output=[id#206L])
+- Exchange hashpartitioning(id#206L, 10), true
   +- *(3) HashAggregate(keys=[id#206L], functions=[], output=[id#206L])
      +- Union
         :- *(1) Range (0, 10, step=1, splits=2)
         +- *(2) Range (0, 10, step=1, splits=2)
```

It's too expensive to detect node removal so we make a compromise only to copy tags to node with no tags.

### Does this PR introduce _any_ user-facing change?
Yes. Fix a bug.

### How was this patch tested?
Add test.

### CODE REVIEW

**Comment by manuzhang** [comment]
cc @cloud-fan @maryannxue

**Comment by cloud-fan** [comment]
Can you explain more in details? Why do we lose the final aggregate?

**Comment by manuzhang** [comment]
https://github.com/apache/spark/blob/701e5934140e477426cf2a68f2c5746be67418d6/sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala#L494-L514

The query stage to replace, `ShuffleQueryStage`, is linked to logical node `Aggregate(...)` instead of `RepartitionByExpression(Aggregate(...))` without update in eliminating Exchange.  That lead to a new physical node of `ShuffleQueryStage` which will replace `HashAggregate(ShuffleQueryStage...)` where the final aggregate is lost

**Comment by cloud-fan** [comment]
> That leads to a new physical node of ShuffleQueryStage which will replace HashAggregate(ShuffleQueryStage...) where the final aggregate is lost

Can you elaborate more? Is the "new physical node" produced by reoptimizing and replanning?

**Comment by maryannxue** [comment]
I think the problem is exactly the other way around, when doing this transform in `EnsureRequirements`, the tags on the final `Aggregate` are overridden by the those on the top `Exchange`, then its logical link has been wrongly changed to `Repartition`. So then in `replaceWithQueryStagesInLogicalPlan`, the stage didn't match the top agg successfully.

**Comment by maryannxue** [comment]
We should prevent tag overriding when copying tags i think.

**Comment by manuzhang** [comment]
https://github.com/apache/spark/blob/2a88a202719839f96838307fc9a8b4dc9bfd0c34/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala#L332-L345

@maryannxue, that makes more sense. So do you mean change the `newNode.copyTagsFrom(this)` here and `transformDown` to

```
if (newNode.isEmpty) {
  newNode.copyTagsFrom(this)
}
```

**Comment by cloud-fan** [comment]
We should change `copyTagsFrom` to
```
    other.tags.foreach{ t =>
      if (!tags.contains(t._1)) {
        tags += t
      }
    }
```

**Comment by cloud-fan** [comment]
The steps lead to this bug are:
1. get logical link of the query stage (Aggregate)
2. get the corresponding physical plan sub-tree (should match final agg, but we match shuffle stage due to the bug)
3. replace Aggregate in the logical plan with logical query stage("partial agg -> shuffle stage"), and get "logical query stage  -> repartition"
4. reoptimize and re-plan, the new physical plan has no final agg

So the key is step 2. We are fine as long as we don't override the logical plan tag in final aggregate with `Repartition`.

**Comment by manuzhang** [comment]
@cloud-fan Thanks for completing the whole story. I've redone the PR as you and @maryannxue suggested

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 1]
```diff
@@ -1205,4 +1205,13 @@ class AdaptiveQueryExecSuite
       })
     }
   }
+
+  test("SPARK-32753: Do not override when copying tags") {
+    withSQLConf(
+      SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> "true"
+    ) {
+      spark.range(10).union(spark.range(10)).createOrReplaceTempView("v1")
+      runAdaptiveAndVerifyResult("SELECT id FROM v1 GROUP BY id DISTRIBUTE BY id")
```
Comment (maryannxue) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:1214]: we can also check the plan here to make sure the repartition did get removed.

**Review by maryannxue — CHANGES_REQUESTED** [review]
I'm wondering if we can take a more conservative approach. Changing `copyTagsFrom` will have a big impact to all tag usages.
Maybe just a trick in `EnsureRequirements` for now by doing `operator.copyTagsFrom(child)`. Means we override the exchange tags first and then the exchange tags will later be copied back to child by the transform method.

**Comment by manuzhang** [comment]
@maryannxue that requires making `copyTagsFrom` a public method. Is it ok ?

**Comment by maryannxue** [comment]
Thats fine. Dont think thats a big deal ;)

**Comment by cloud-fan** [comment]
> that requires making copyTagsFrom a public method. Is it ok ?

It's a bit hacky. Can we unset and set the logical link tag?

**Comment by manuzhang** [comment]
> Maybe just a trick in EnsureRequirements for now by doing `operator.copyTagsFrom(child)`.

This doesn't seem to work as the final `Aggregate` still links to `Repartition` and fails the test.

**Comment by manuzhang** [comment]
> Maybe just a trick in EnsureRequirements for now by doing `operator.copyTagsFrom(child)`

This fails test as `Aggregate` still links to `Repartition`

> Can we unset and set the logical link tag?
Do you mean `operator.setTagValue(SparkPlan.LOGICAL_PLAN_TAG, child.getTagValue(SparkPlan.LOGICAL_PLAN_TAG))` ? This doesn't compile.

**Comment by cloud-fan** [comment]
> Do you mean operator.setTagValue(SparkPlan.LOGICAL_PLAN_TAG, child.getTagValue(SparkPlan.LOGICAL_PLAN_TAG).get) ?

Yea, probably set the TEMP_LOGICAL_PLAN_TAG as well

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala [hunk 2]
```diff
@@ -91,7 +91,11 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product {
   private val tags: mutable.Map[TreeNodeTag[_], Any] = mutable.Map.empty
 
   protected def copyTagsFrom(other: BaseType): Unit = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala:93]: thinking about it more, I think we missed one case of tree node tags. If the rule replaces one node with a new one, it makes sense to copy the tags. If the rule removes one or more nodes, it doesn't make sense to append the tags to an existing node.

However, it's too expensive to detect the removal action (you need to traverse the tree and see if the new node is a descendant of the original node). I think a compromise is
```
if (tags.isEmpty) {
  tags ++= other.tags
}
```

It's still possible that we copy tags to an existing node due to a node removal rule, but it's not a big issue if that node has no tags at all.

Comment (manuzhang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala:93]: looks good. @maryannxue any concerns with this approach ?

Comment (maryannxue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala:93]: Agree with @cloud-fan 's reasoning and proposed fix.

The only worry is that we're not 100% sure if we are breaking things somewhere else, but the tests don't say so yet. Can't think of an approach that would have less impact yet fix this problem perfectly. Let's go with this approach then.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala [hunk 3]
```diff
@@ -91,7 +91,9 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product {
...
-    tags ++= other.tags
+    if (tags.isEmpty) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala:94]: can we add some comments?

Comment (manuzhang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala:97]: done

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by cloud-fan** [comment]
@manuzhang can you open a new PR for 3.0?

**Comment by manuzhang** [comment]
@cloud-fan Sure, thanks for review

**Pull request closed without merging**